go/libraries/doltcore/merge/violations_fk_prolly.go

```diff
@@ -146,39 +146,55 @@ func prollyChildSecDiffFkConstraintViolations(
 	ctx context.Context,
 	foreignKey doltdb.ForeignKey,
 	postParent, postChild *constraintViolationsLoadedTable,
-	preChildScndryIdx prolly.Map,
+	preChildSecIdx prolly.Map,
 	receiver FKViolationReceiver) error {
 	postChildRowData := durable.ProllyMapFromIndex(postChild.RowData)
-	postChildScndryIdx := durable.ProllyMapFromIndex(postChild.IndexData)
-	parentScndryIdx := durable.ProllyMapFromIndex(postParent.IndexData)
-
-	idxDesc, _ := parentScndryIdx.Descriptors()
-	partialDesc := idxDesc.PrefixDesc(len(foreignKey.TableColumns))
-	partialKB := val.NewTupleBuilder(partialDesc)
+	postChildSecIdx := durable.ProllyMapFromIndex(postChild.IndexData)
+	parentSecIdx := durable.ProllyMapFromIndex(postParent.IndexData)
 
-	primaryKD, _ := postChildRowData.Descriptors()
-	kb := val.NewTupleBuilder(primaryKD)
+	parentSecIdxDesc, _ := parentSecIdx.Descriptors()
+	partialDesc := parentSecIdxDesc.PrefixDesc(len(foreignKey.TableColumns))
+	childPriKD, _ := postChildRowData.Descriptors()
+	childPriKB := val.NewTupleBuilder(childPriKD)
 
-	err := prolly.DiffMaps(ctx, preChildScndryIdx, postChildScndryIdx, func(ctx context.Context, diff tree.Diff) error {
+	var parentSecIdxCur *tree.Cursor
+	err := prolly.DiffMaps(ctx, preChildSecIdx, postChildSecIdx, func(ctx context.Context, diff tree.Diff) error {
 		switch diff.Type {
 		case tree.AddedDiff, tree.ModifiedDiff:
-			k, v := val.Tuple(diff.Key), val.Tuple(diff.To)
-			partialKey, hasNulls := makePartialKey(
-				partialKB,
-				foreignKey.TableColumns,
-				postChild.Index,
-				postChild.IndexSchema,
-				k,
-				v,
-				preChildScndryIdx.Pool())
-			if hasNulls {
-				return nil
+			k := val.Tuple(diff.Key)
+			// TODO: possible to skip this if there are not null constraints over entire index
+			for i := 0; i < k.Count(); i++ {
+				if k.FieldIsNull(i) {
+					return nil
+				}
 			}
 
-			err := createCVIfNoPartialKeyMatchesSec(ctx, k, v, partialKey, partialDesc, primaryKD, kb, parentScndryIdx, postChildRowData, postChildRowData.Pool(), receiver)
+			if parentSecIdxCur == nil {
+				newCur, err := tree.NewCursorAtKey(ctx, parentSecIdx.NodeStore(), parentSecIdx.Node(), k, partialDesc)
+				if err != nil {
+					return err
+				}
+				if !newCur.Valid() {
+					return createNewCVForSecIdx(ctx, k, childPriKD, childPriKB, postChildRowData, postChildRowData.Pool(), receiver)
+				}
+				parentSecIdxCur = newCur
+			}
+
+			err := tree.Seek(ctx, parentSecIdxCur, k, partialDesc)
 			if err != nil {
 				return err
 			}
+			if !parentSecIdxCur.Valid() {
+				return createNewCVForSecIdx(ctx, k, childPriKD, childPriKB, postChildRowData, postChildRowData.Pool(), receiver)
+			}
+
+			rng := prolly.PrefixRange(k, partialDesc)
+			key := val.Tuple(parentSecIdxCur.CurrentKey())
+			if !rng.Matches(key) {
```

Comment:
this one is over my head. maybe add a comment here to indicate which cases require the extra range match, otherwise we miss constraint violations?

Reply:
The range check is still necessary here because it's possible that the key in question `k` is smaller than every key in `parentSecIdxCur`, so the cursor is still valid.

Reply:
would it be equivalent to do something like `paretnSecIdx.Tuples().Order.Compare(k, parentSecIdxCur.CurrentKey())`? might be more self-documenting